### PR TITLE
ENH: Enable ExcelWriter to construct in-memory sheets

### LINKED ISSUE #6403: ExcelFile parse of empty sheet fails with "IndexError: list index out of range"
Using pandas 0.13.1 on OS X Mavericks to parse a blank Excel spreadsheet causes "IndexError: list index out of range". Apparently the default header=0 in ```_parse_excel``` causes the execution of ```_trim_excel_header(data[header])```. Perhaps when nrows==0 this should not be executed.

```python
import pandas as pd
xl_file = pd.ExcelFile('blank.xlsx')
xl_file.parse('Sheet1') #Sheet1 has no data
```

STDERR:

```
Traceback (most recent call last):
  File "/Users/myourshaw/lab/pypeline/python2/excel_example.py", line 10, in <module>
    xl_file.parse('Sheet1')
  File "/usr/local/Cellar/python/2.7.6/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/io/excel.py", line 208, in parse
    **kwds)
  File "/usr/local/Cellar/python/2.7.6/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/io/excel.py", line 291, in _parse_excel
    data[header] = _trim_excel_header(data[header])
IndexError: list index out of range
```

### LINKED ISSUE #7074: ExcelWriter does not support BytesIO input
At least XlsxWriter supports writing to buffers but the `ExcelWriter` constructors try to read the extension of the supplied `path`, so it fails when you initialise it as `ExcelWriter(buf, engine="xlsxwriter")`.

Currently you can hack that into pandas using

    b = BytesIO()
    w = ExcelWriter("data.xlsx", engine="xlsxwriter")
    w.book = xlsxwriter.Workbook(b)
    df.to_excel(w)
    w.save()
    b.getvalue()

but it would of course be nice if this just worked out-of-the-box. I'll try to build a PR for that.

### LINKED ISSUE #7171: PERF: vbench for excel writers
head to head for all the writer engines

``openpyxl``, ``xlsxwriter``

### LINKED ISSUE #8188: Suggested improvement: allow ExcelWriter to automatically convert lists and dict to strings
Problem description:
I have pandas dataframes with diverse data types and structures, including lists. The csv-writer saves the lists without trouble, but I prefer to also save in excel format for collaboration with my colleagues.
Currently the ExcelWriter returns a TypeError if the dataframe contains a list.
`TypeError: float() argument must be a string or a number, not 'list'`

It would be really great if ExcelWriter accepted lists and dictionaries and tuples.
Would it cause any problem to simply convert these data formats to strings before saving?
Sorry I’m not an experienced programmer, so I haven‘t made a pull request and attempted to improve the module myself.
Here is some code (Python 3.4, pandas 0.14.0) to describe the current situation.
```
import pandas as pd
#create a new dataframe that includes a list and a dict
title = pd.Series(['Toy Story (1995)',
 'Jumanji (1995)',
 'Grumpier Old Men (1995)',
 'Waiting to Exhale (1995)',
 'Father of the Bride Part II (1995)'], index = [0 ,1, 2, 3, 4], name='titles')
genre_as_list = pd.Series([['Animation', "Children's", 'Comedy'],
 ['Adventure', "Children's", 'Fantasy'],
 ['Comedy', 'Romance'],
 ['Comedy', 'Drama'],
 ['Comedy']], index = [0 ,1, 2, 3, 4], name='genre_as_list')
rating_as_dict = pd.Series([{'F':0.5,'M':0.6},
 {'F':0.1,'M':0.2},
 {'F':0.3,'M':0.3},
 {'F':0.8,'M':0.8},
 {'F':0.8,'M':0.6}], index = [0 ,1, 2, 3, 4], name='rating_as_dict')
df = pd.concat([title, genre_as_list, rating_as_dict], axis=1)
```
attempt to save the dataframe in excel
```
writer = pd.ExcelWriter('saved_df.xlsx')
df.to_excel(writer, sheet_name='Sheet1')
```
which will result in
`TypeError: float() argument must be a string or a number, not 'list'`
to save the data, it is necessary to convert the python lists or dicts to strings, separated by ", " or pipes "|"
```
df['genre_as_string'] = [str(i) for i in df['genre_as_list']]
df['rating_as_string'] = [str(i) for i in df['rating_as_dict']]
df['genre_pipes_string'] = ['|'.join(i) for i in df['genre_as_list']]
#in my ipython output, the difference is not actually visible between the lists and the stringlists
print('first item from list: %s\nfirst item from stringlist: %s' % (df['genre_as_list'][0][0],
                                                          df['genre_as_string'][0][0]))
```
After deleting any columns containing a python list or dict, I can now save without any problem.
```
df = df.drop(['genre_as_list','rating_as_dict'], axis=1)
writer = pd.ExcelWriter('saved_df.xlsx')
df.to_excel(writer, sheet_name='Sheet1')
writer.save()
```
If I load from the saved excel sheet, I can confirm that the data is intact
```
df_from_excel = pd.read_excel('saved_df.xlsx', sheetname='Sheet1')
#convert back to original python lists and dict, if desired
df_from_excel['genre_as_list_again'] = [eval(i) for i in df_from_excel['genre_as_string']]
df_from_excel['rating_as_dict_again'] = [eval(i) for i in df_from_excel['rating_as_string']]
df_from_excel.loc[0,'genre_as_list_again'][0]
```

### PR DESCRIPTION
**Pull request by bashtage**
Add support for StringIO/BytesIO to ExcelWriter
Add vbench support for writing excel files
Add support for serializing lists/dicts to strings
Fix bug when reading blank excel sheets

closes #8188
closes #7074
closes #6403
closes #7171

### CODE REVIEW

File: doc/source/install.rst [hunk 1]
```diff
@@ -249,10 +249,9 @@ Optional Dependencies
 * `statsmodels <http://statsmodels.sourceforge.net/>`__
    * Needed for parts of :mod:`pandas.stats`
 * `openpyxl <http://packages.python.org/openpyxl/>`__, `xlrd/xlwt <http://www.python-excel.org/>`__
-   * openpyxl version 1.6.1 or higher, but lower than 2.0.0
```
Comment (jreback) [inline doc/source/install.rst]: did you mean to take this out?

File: pandas/io/excel.py [hunk 2]
```diff
@@ -523,7 +530,11 @@ def __new__(cls, path, engine=None, **kwargs):
         # only switch class if generic(ExcelWriter)
         if cls == ExcelWriter:
```
Comment (jreback) [inline pandas/io/excel.py]: use ``isinstance(cls, ExcelWriter)`` (I know you didn't touch that code...)

File: pandas/io/tests/test_excel.py [hunk 3]
```diff
@@ -1218,6 +1241,25 @@ def test_datetimes(self):
 
             tm.assert_series_equal(write_frame['A'], read_frame['A'])
 
+    # GH7074
+    def test_bytes_io(self):
+        bio = BytesIO()
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: awesome. can you add a small section to the excel docs (io.rst) showing writing to bytes?

File: pandas/io/excel.py [hunk 4]
```diff
@@ -417,10 +419,13 @@ def _parse_cell(cell_contents,cell_typ):
                     if parse_cols is None or should_parse[j]:
                         row.append(_parse_cell(value,typ))
                 data.append(row)
-    
+
+            if sheet.nrows == 0:
+                return DataFrame()
+
             if header is not None:
```
Comment (jreback) [inline pandas/io/excel.py]: what about ``sheet.nrows==1`` (and its the header)? do we construct an empty frame with the correct columns?

File: pandas/io/tests/test_excel.py [hunk 5]
```diff
@@ -520,6 +520,29 @@ def test_reader_seconds(self):
...
+        _skip_if_no_xlrd()
+
+        blank = os.path.join(self.dirpath, 'blank.xls')
+        actual = read_excel(blank, 'Sheet1')
+        tm.assert_frame_equal(actual, DataFrame())
+
+        blank = os.path.join(self.dirpath, 'blank.xlsx')
+        actual = read_excel(blank, 'Sheet1')
+        tm.assert_frame_equal(actual, DataFrame())
+
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: can you also tests a round-trip of writing (with no header), then reading?

File: pandas/io/tests/test_excel.py [hunk 6]
```diff
@@ -1218,6 +1241,25 @@ def test_datetimes(self):
...
+        df = DataFrame(np.random.randn(10, 2))
+        writer = ExcelWriter(bio)
+        df.to_excel(writer)
+        writer.save()
+        bio.seek(0)
+        reread_df = pd.read_excel(bio)
+        tm.assert_frame_equal(df, reread_df)
+
+    def test_write_lists_dict(self):
+        df = pd.DataFrame({'mixed': ['a', ['b', 'c'], {'d': 'e', 'f': 2}],
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: pls add the issue number

File: pandas/io/tests/test_excel.py [hunk 7]
```diff
@@ -1218,6 +1241,25 @@ def test_datetimes(self):
...
+                           'numeric': [1, 2, 3.0],
+                           'str': ['apple', 'banana', 'cherry']})
+        with ensure_clean(self.ext) as path:
+            df.to_excel(path, 'Sheet1')
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: I think we should document this, doc-string?

File: vb_suite/packers.py [hunk 8]
```diff
@@ -206,3 +207,25 @@ def remove(f):
...
 packers_write_stata_with_validation = Benchmark("df.to_stata(f, {'index': 'tc'})", setup, cleanup="remove(f)", start_date=start_date)
+
+#----------------------------------------------------------------------
+# Excel - alternative writers
+setup = common_setup + """
+bio = BytesIO()
+"""
+
+benchmark_xlwt = """
+writer = pd.io.excel.ExcelWriter(bio, engine='xlsxwriter')
```
Comment (jreback) [inline vb_suite/packers.py]: can you add the reading as well?

File: pandas/io/excel.py (hunk 4 above)
Comment (bashtage) [inline pandas/io/excel.py]: Yes, I added a test for this case and it works correctly already.

File: doc/source/install.rst (hunk 1 above)
Comment (bashtage) [inline doc/source/install.rst]: Yes - both 1.x and 2.x are clearly supported and tested in excel.py and test_excel.py.  I suppose the person who added the OpenPyXL 2.x path forgot to change this.

File: pandas/io/excel.py (hunk 4 above)
Comment (jreback) [inline pandas/io/excel.py]: yep, saw that below. thanks.

File: doc/source/install.rst (hunk 1 above)
Comment (jreback) [inline doc/source/install.rst]: hmm, IIRC there were some issues with >=2.0.0 and < 2.0.3 something like that (e.g. some style sheet issues. But ok, past that now, so ok.

File: pandas/io/tests/test_excel.py (hunk 3 above)
Comment (bashtage) [inline pandas/io/tests/test_excel.py]: I added something

File: pandas/io/tests/test_excel.py (hunk 7 above)
Comment (bashtage) [inline pandas/io/tests/test_excel.py]: I added a line

File: vb_suite/packers.py (hunk 8 above)
Comment (bashtage) [inline vb_suite/packers.py]: Added

File: doc/source/io.rst [hunk 9]
```diff
@@ -2184,6 +2184,38 @@ argument to ``to_excel`` and to ``ExcelWriter``. The built-in engines are:
 
 .. _io.clipboard:
```
Comment (jreback) [inline doc/source/io.rst]: move your doc to before this refernce (for the next section)

File: doc/source/io.rst [hunk 10]
```diff
@@ -2184,6 +2184,38 @@ argument to ``to_excel`` and to ``ExcelWriter``. The built-in engines are:
...
 
+Writing Excel Files to Memory
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Pandas supports writing Excel files to buffer-like objects such as StringIO or
+BytesIO using :class:`~pandas.io.excel.ExcelWriter`.
```
Comment (jreback) [inline doc/source/io.rst]: back ticks around StringIO/BytesIO

File: doc/source/io.rst [hunk 11]
```diff
@@ -2184,6 +2184,38 @@ argument to ``to_excel`` and to ``ExcelWriter``. The built-in engines are:
...
+   df.to_excel(writer, sheet_name='Sheet1')
+
+   # Save the workbook
+   writer.save()
+
+   # Seek to the beginning and read to copy the workbook to a variable in memory
+   bio.seek(0)
+   workbook = bio.read()
+
+.. note:: `engine` is optional but recommended.  Setting the engine determines
```
Comment (jreback) [inline doc/source/io.rst]: I think that the text has to start on the next line

File: doc/source/io.rst [hunk 12]
```diff
@@ -2184,6 +2184,38 @@ argument to ``to_excel`` and to ``ExcelWriter``. The built-in engines are:
...
+    the version of workbook produced. Setting `engine='xlrd'` will produce an
+    Excel 2003-format workbook (xls).  Using either `'openpyxl'` or
```
Comment (jreback) [inline doc/source/io.rst]: double-backticks around the ``engine='xlrd'`` (and openpyxl) etc

File: doc/source/whatsnew/v0.17.0.txt [hunk 13]
```diff
@@ -53,9 +55,13 @@ Removal of prior version deprecations/changes
 
 Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
+- Added vbench benchmarks for Excel writers (:issue:`7171`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: readers/writers

File: pandas/io/tests/test_excel.py [hunk 14]
```diff
@@ -1218,6 +1241,26 @@ def test_datetimes(self):
...
+
+    # GH8188
+    def test_write_lists_dict(self):
+        df = pd.DataFrame({'mixed': ['a', ['b', 'c'], {'d': 'e', 'f': 2}],
+                           'numeric': [1, 2, 3.0],
+                           'str': ['apple', 'banana', 'cherry']})
+        with ensure_clean(self.ext) as path:
+            df.to_excel(path, 'Sheet1')
+            read = read_excel(path, 'Sheet1', header=0)
+
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: can you assert someting about the returned value here (best if you can construct the actual returned frame, but at the very least assert the columns / dtypes)

**Comment by bashtage** [comment]
@jreback The vbench for openpyxl is very slow ~ around 300s on my computer. This is around 10x slower than the next slowest test (sql).  Should I reduce the df size for this test?

**Comment by jreback** [comment]
yes, a vbench should be < 1s if at all possible, so reduce the size on these.

**Comment by bashtage** [comment]
I think it is finished once green - I snuck in one more Excel-related issue (testing xlwt on Py34).

Openpyxl is very slow compared to xlsxwriter (5x slower)

**Comment by jreback** [comment]
ahh, the ``xlwt`` just got ported to py3. finally! excellent.

yes, ``openpyxl`` does have this perf issue (and why ``xlsxwriter`` is the default). Though the author is working on making it much better.

ping when green.

**Comment by bashtage** [comment]
@jreback ready

File: doc/source/io.rst [hunk 15]
```diff
@@ -2182,6 +2184,40 @@ argument to ``to_excel`` and to ``ExcelWriter``. The built-in engines are:
 
    df.to_excel('path_to_file.xlsx', sheet_name='Sheet1')
```
Comment (jreback) [inline doc/source/io.rst]: can you add a reference here (so we can link to it)

File: doc/source/whatsnew/v0.17.0.txt [hunk 16]
```diff
@@ -38,6 +38,8 @@ Backwards incompatible API changes
 
 Other API Changes
 ^^^^^^^^^^^^^^^^^
+- Enable writing Excel files in memory using StringIO/BytesIO (:issue:`7074`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: link the above doc reference

**Comment by jreback** [comment]
minor doc comment. ping after you have pushed for merging.

**Comment by bashtage** [comment]
@jreback Added those.

**Comment by bashtage** [comment]
Rebased

**Comment by jreback** [comment]
@bashtage thanks for this!

**Pull request merged**